[docs] Decorator to create a deprecation warning

PaliC opened the pull request:
This PR adds the `@deprecate` decorator for internal functions which we are prepping for deprecation.  Add it on top of an internal function to emit a deprecation warning + allow bc with the non internal version of the function.

Tested with `python test/test_utils.py TestDeprecate.test_deprecated `

Furthermore, testing with a modified version of the tes in the pr gives something like this which is what we want

```
/home/sahanp/repos/pytorch/test/test_utils.py:1239: UserWarning: deprecated_api is DEPRECATED, please consider using an alternative API(s). 
  deprecated_api(1, 2)
  ```

PaliC committed:
add deprecation decorator

Skylion007 commented on torch/_utils_internal.py:
```diff
@@ -282,3 +283,62 @@ def record_chromium_event_internal(
     return True
+
+
+def deprecate(version: str):
```
Ugh, why not just extend typing_extensions.deprecated for this?

Skylion007 commented on torch/_utils_internal.py:
```diff
@@ -282,3 +283,62 @@ def record_chromium_event_internal(
+
+        # public alias
+        @functools.wraps(func)
+        def alias(*args: _P.args, **kwargs: _P.kwargs):  # type: ignore[misc]
```
```suggestion
        def alias(*args: _P.args, **kwargs: _P.kwargs) -> _T:

Skylion007 replied:
Same, using typing_extensions.deprecated here as the wrapper

Skylion007 replied:
Why are we dropping the return type though?

Skylion007 commented on torch/_utils_internal.py:
```diff
@@ -282,3 +283,62 @@ def record_chromium_event_internal(
+        setattr(module, public_name, alias)
+
+        @functools.wraps(func)
+        def wrapper(*args: _P.args, **kwargs: _P.kwargs):  # type: ignore[misc]
```
```suggestion
        def wrapper(*args: _P.args, **kwargs: _P.kwargs) -> T:

Skylion007 replied:
Just typing_extensions.deprecated here as the wrapper (with the message filled in)

Skylion007 replied:
Why not propagate it's return type?

Skylion007 commented on torch/_utils_internal.py:
```diff
@@ -282,3 +283,62 @@ def record_chromium_event_internal(
+        # public alias
+        @functools.wraps(func)
+        def alias(*args: _P.args, **kwargs: _P.kwargs):  # type: ignore[misc]
+            warnings.warn(warning_msg, DeprecationWarning, stacklevel=2)
```
Or just use https://typing-extensions.readthedocs.io/en/latest/#typing_extensions.deprecated

PaliC committed:
feedback

Skylion007 commented on torch/_utils_internal.py:
```diff
@@ -314,14 +311,12 @@ def decorator(func: Callable[_P, _T]) -> Callable[_P, _T]:
         @functools.wraps(func)
         def alias(*args: _P.args, **kwargs: _P.kwargs):  # type: ignore[misc]
-            warnings.warn(warning_msg, DeprecationWarning, stacklevel=2)
+            warnings.warn(warning_msg, UserWarning, stacklevel=2)
```
```suggestion
            warnings.warn(warning_msg, FutureWarning, stacklevel=2)

PaliC committed:
feedback

Skylion007 commented on torch/_utils_internal.py:
```diff
@@ -282,3 +283,56 @@ def record_chromium_event_internal(
+
+        # public deprecated alias
+        @functools.wraps(func)
+        def alias(*args: _P.args, **kwargs: _P.kwargs):  # type: ignore[misc]
+            warnings.warn(warning_msg, UserWarning, stacklevel=2)
+            return func(*args, **kwargs)
```
```suggestion
        alias = typing_extensions.deprecated(warning_message, category=UserWarning, stacklevel=2)
```

PaliC committed:
feedback

PaliC committed:
better docstring

PaliC committed:
better docstring

PaliC committed:
Update torch/_utils_internal.py

Co-authored-by: albanD <desmaison.alban@gmail.com>

PaliC committed:
feedback

PaliC committed:
Update torch/_utils_internal.py

Co-authored-by: albanD <desmaison.alban@gmail.com>

albanD reviewed (approved):
Optional style nit, sgtm otherwise once lint is fixed

PaliC commented:
@pytorchbot rebase

PaliC committed:
Update torch/_utils_internal.py

Co-authored-by: albanD <desmaison.albangmail.com>

ghstack-source-id: 3adc027f56d99d900d9b364ed74d2ab5ce6f6460
Pull Request resolved: https://github.com/pytorch/pytorch/pull/156829
